coil-core/src/androidMain/kotlin/coil3/util/NetworkObserver.kt

```diff
@@ -50,7 +53,12 @@ private class RealNetworkObserver(
         val request = NetworkRequest.Builder()
             .addCapability(NET_CAPABILITY_INTERNET)
             .build()
-        connectivityManager.registerNetworkCallback(request, networkCallback)
```

Comment:
We don't have to also change `val isAnyOnline = connectivityManager.allNetworks.any` below to only consider the default network?

Reply:
Good question - it doesn't look like it.

We had been asked by Google to only make the `registerNetworkCallback ` to `registerDefaultNetworkCallback` change in our application, and we could then verify network requests using `db shell dumpsys connectivity | grep -i package.name`. If we got network logs of type TRACK_DEFAULT it meant everything was good and using the correct network to make the requests, but if we received LISTEN type logs it was using the incorrect network. I could verify with our test app that after the change I made, we now only got TRACK_DEFAULT logs when using coil to load images. 

I can share more info on how to reproduce on an automotive emulator if you're interested, as it's not straight forward!

Reply:
Hi, James here. I'm an Android Platform lead at Google for the Connectivity teams. Thanks Danielle for mentioning this change. (and putting it up). I wanted to provide some input in case it is helpful.

Regarding Network.isOnline, what is the purpose of that code? I would assume, it checks to see if any network is online by virtue of naming. However, it doesn't seem to have any connection to the passed in `networkCallback` in this commented block of code? I see the callback triggers `isOnline`, however whatever network was passed back via the callback is then ignored by usage of `connectivityManager.allNetworks`.

Given that, what question is `isOnline` trying to answer? Here are the two questions I would guess:

- Is there a network available to me in my context? I.e., is there a network available I can use.
- Is there any network available at all whether I have access to it or not? I.e., there could be a VPN available with `INTERNET` capabilities that I don't have access to and can't use, but I would still want to show `isOnline` as `true`.

I'm guessing it is the former. Is there a network available that I can use? If so, I will recommend additional changes, but want to confirm before doing so.

Reply:
@tech31 Yep, you're right that `isOnline` (and more broadly `NetworkObserver`) is trying to determine if there's an available network with internet that we can use. Though, reading your other comment I think we probably only want to check the default network by default. Essentially we want to know if `OkHttpClient.newCall`/`HttpUrlConnection`/`HttpClient.request` will fail immediately. If so, we use this info to skip querying the network and read from the disk cache.

Reply:
Ok cool. Yes, in this case, you should always rely on the default network for the reasons I mentioned above (such as being put on a VPN, etc). Otherwise what you are returned may/will not accurately cover the intended question "do I have access to the internet".

Reply:
@tech31 How should we check if the default network has `NET_CAPABILITY_INTERNET`? `registerDefaultNetworkCallback` doesn't accept a `NetworkRequest` and `NetworkCallback`'s methods all say `Do NOT call {@link #getNetworkCapabilities(Network)} or other synchronous ConnectivityManager methods in this callback as this is prone to race conditions`, which `Network.isOnline` relies on. It's not clear how to check the default network's capabilities.